autogen/oai/together.py

```diff
@@ -0,0 +1,369 @@
+
+Resources:
+- https://docs.together.ai/docs/inference-python
+"""
+
+from __future__ import annotations
+
+import base64
+import copy
+import os
+import random
+import re
+import time
+import warnings
+from io import BytesIO
+from typing import Any, Dict, List, Mapping, Tuple, Union
+
+import requests
+from openai.types.chat import ChatCompletion, ChatCompletionMessageToolCall
+from openai.types.chat.chat_completion import ChatCompletionMessage, Choice
+from openai.types.completion_usage import CompletionUsage
+from PIL import Image
+
+# pip install together
+from together import Together, error
+
+from autogen.oai.client_utils import should_hide_tools, validate_parameter
+
+
+class TogetherClient:
+    """Client for Together.AI's API."""
+
+    def __init__(self, **kwargs):
+        """Requires api_key or environment variable to be set
+
+        Args:
+            api_key (str): The API key for using Together.AI (or environment variable TOGETHER_API_KEY needs to be set)
+        """
+        # Ensure we have the api_key upon instantiation
+        self.api_key = kwargs.get("api_key", None)
+        if not self.api_key:
+            self.api_key = os.getenv("TOGETHER_API_KEY")
+
+        assert (
+            self.api_key
+        ), "Please include the api_key in your config list entry for Together.AI or set the TOGETHER_API_KEY env variable."
+
+    def message_retrieval(self, response) -> List:
+        """
+        Retrieve and return a list of strings or a list of Choice.Message from the response.
+
+        NOTE: if a list of Choice.Message is returned, it currently needs to contain the fields of OpenAI's ChatCompletion Message object,
+        since that is expected for function or tool calling in the rest of the codebase at the moment, unless a custom agent is being used.
+        """
+        return [choice.message for choice in response.choices]
+
+    def cost(self, response) -> float:
+        return response.cost
+
+    @staticmethod
+    def get_usage(response) -> Dict:
+        """Return usage summary of the response using RESPONSE_USAGE_KEYS."""
+        # ...  # pragma: no cover
+        return {
+            "prompt_tokens": response.usage.prompt_tokens,
+            "completion_tokens": response.usage.completion_tokens,
+            "total_tokens": response.usage.total_tokens,
+            "cost": response.cost,
+            "model": response.model,
+        }
+
+    def parse_params(self, params: Dict[str, Any]) -> Dict[str, Any]:
+        """Loads the parameters for Together.AI API from the passed in parameters and returns a validated set. Checks types, ranges, and sets defaults"""
+        together_params = {}
+
+        # Check that we have what we need to use Together.AI's API
+        together_params["model"] = params.get("model", None)
+        assert together_params[
+            "model"
+        ], "Please specify the 'model' in your config list entry to nominate the Together.AI model to use."
```

Comment:
We should assume all `create` is called from OpenAIWrapper, and we already checked that model params is passed there. (We need to double check that, if not, update it there)

If anything, we should check that model is in togetherai. I don't think it is plausible to maintain this, unless we check together ai everyday. So I think the best way here is do nothing, let togetherai check whether the model is in their API.

Reply:
Thanks for running through this @yiranwu0, let me work through the Mistral review and then come back to this.

Reply:
Okay, I'll remove the model check and just put the model in the parameters for use for the create call.

Reply:
> We should assume all `create` is called from OpenAIWrapper, and we already checked that model params is passed there. (We need to double check that, if not, update it there)

If I exclude the 'model' from my config it gets into the create function without an error/exception.

I'll leave the check in the code for now (as it's also in Anthropic, Mistral, and Gemini checks it as well). However, happy to take it out if we're handling it further up the chain.